src/libstd/collections/hash/map.rs

```diff
@@ -1711,18 +1657,26 @@ impl<K, S, Q: ?Sized> super::Recover<Q> for HashMap<K, (), S>
             return None
         }
 
-        self.search_mut(key).map(|bucket| pop_internal(bucket).0)
+        self.search_mut(key).into_occupied_bucket().map(|bucket| pop_internal(bucket).0)
     }
 
     fn replace(&mut self, key: K) -> Option<K> {
         let hash = self.make_hash(&key);
         self.reserve(1);
 
-        let mut retkey = None;
-        self.insert_or_replace_with(hash, key, (), |key_ref, _, key, _| {
-            retkey = Some(replace(key_ref, key));
-        });
-        retkey
+        match search_hashed(&mut self.table, hash, |k| *k == key) {
+            InternalEntry::Occupied { mut elem } => {
+                Some(mem::replace(elem.read_mut().0, key))
+            }
+            other => {
+                if let Some(Vacant(vacant)) = other.into_entry(key) {
```

Comment:
Is there any reason to not just fully match on the `InternalEntry` and avoid the `if let`? Admittedly, the entry creation would need to be duplicated, but it seems like a net win to me.